plugins/py3embed/py3embed.cpp

```diff
@@ -431,15 +432,30 @@ static class Python3xGlobalState
 protected:
     static StringBuffer &wrapPythonText(StringBuffer &out, const char *in, const char *params)
     {
-        out.appendf("def __user__(%s):\n  ", params);
-        char c;
-        while ((c = *in++) != '\0')
+        // Complicated by needing to keep future import lines outside defined function
+        // Per python spec, a future statement must appear near the top of the module. The only lines that can appear before a future statement are:
+        //   the module docstring (if any),
+        //   comments,
+        //   blank lines, and
+        //   other future statements.
+        // We don't attempt to parse the python to spot these - instead, we pull all lines up to and including the last future statement out to the global scope.
+        // Because this is a little unsophisticated it will be fooled by code that includes things that look like future statements inside multiline strings.
+        // I don't care.
+        StringArray lines;
+        lines.appendList(in, "\n", false);
+        RegExpr expr("^ *from +__future__ +import ");
```

Comment:
Should this be a static object?  It is relatively expensive to set up.

Reply:
I wondered, but then there would be thread safety concerns. 

From jregexp.hpp:

// WARNING - this is not threadsafe - it is up to the caller to ensure that a RegExpr is accessed from only one thread at a time 
// that INCLUDES the simple find() method